src/app/server-cluster/ServerClusterInterfaceRegistry.h

```diff
@@ -0,0 +1,142 @@
+#include <app/server-cluster/ServerClusterInterface.h>
+#include <lib/core/CHIPError.h>
+#include <lib/core/DataModelTypes.h>
+
+namespace chip {
+namespace app {
+
+/// Allows registering and retrieving ServerClusterInterface instances for specific cluster paths.
+class ServerClusterInterfaceRegistry
+{
+public:
+    /// represents an iterable list of clusters
+    class ClustersList
+    {
+    public:
+        class Iterator
+        {
+        public:
+            Iterator(ServerClusterInterface * interface) : mInterface(interface) {}
+
+            Iterator & operator++()
+            {
+                if (mInterface != nullptr)
+                {
+                    mInterface = mInterface->GetNextListItem();
+                }
+                return *this;
+            }
+            bool operator==(const Iterator & other) const { return mInterface == other.mInterface; }
+            bool operator!=(const Iterator & other) const { return mInterface != other.mInterface; }
+            ServerClusterInterface * operator*() { return mInterface; }
+
+        private:
+            ServerClusterInterface * mInterface;
+        };
+
+        ClustersList(ServerClusterInterface * start) : mStart(start) {}
+        Iterator begin() { return mStart; }
+        Iterator end() { return nullptr; }
+
+    private:
+        ServerClusterInterface * mStart;
+    };
+
+    ~ServerClusterInterfaceRegistry();
+
+    /// Associate a specific interface with the given endpoint.
+    ///
+    /// A `ServerClusterInterface` may only be associated on a single endpointId and
+    /// there can be only a single registration for a given `endpointId/clusterId` path.
+    ///
+    /// This means Register WILL RETURN AN ERROR if:
+    ///   - A registration on the given `endpointId/cluster->GetClusterID()` already exists
+    ///   - The given `cluster` is already registered for some enpoint id
+    ///
+    /// Registrations need a valid endpointId and cluster MUST return a valid cluster id.
+    [[nodiscard]] CHIP_ERROR Register(EndpointId endpointId, ServerClusterInterface * cluster);
+
+    /// Remove an existing registration for a given endpoint/cluster path.
+    ///
+    /// Returns the previous registration if any exists (or nullptr if nothing
+    /// to unregister)
+    ServerClusterInterface * Unregister(const ConcreteClusterPath & path);
+
+    /// Return the interface registered for the given cluster path or nullptr if one does not exist
+    ServerClusterInterface * Get(const ConcreteClusterPath & path);
+
+    /// ClustersList is valid as a snapshot only and should not be saved.
+    ClustersList ClustersOnEndpoint(EndpointId endpointId);
+
+    /// Unregister all registrations for the given endpoint.
+    void UnregisterAllFromEndpoint(EndpointId endpointId);
+
+    /// Access to an application global registry of server cluster interfaces
+    /// that are registered within the application.
+    static ServerClusterInterfaceRegistry & Instance();
```

Comment:
Some other singleton should be the one to return this non-singleton, especially to allow moving away from having singletons at all in the process. This singleton single-handedly now makes it impossible to host 2 device data models in 1 process, for no reason.

Reply:
This would assume that this is our main way of consuming these interfaces.

However I am  claiming this is only the "ember consumption point" and ember will be singleton.

I plan for the code-backed provider to be something else, in particular that something else needs to manage endpoint metadata and client clusters anyway, so I specifically do not plan to re-use this registry. This registry is for ember only.